lib/Dialect/FIRRTL/Transforms/InferWidths.cpp

```diff
@@ -402,14 +418,306 @@ static void solveBinary(BinaryExpr *expr,
+        if (ineq.rec_scale != 1)
+          note << ineq.rec_scale;
+        note << "W";
+        if (ineq.rec_bias < 0)
+          note << "-" << -ineq.rec_bias;
+        if (ineq.rec_bias > 0)
+          note << "+" << ineq.rec_bias;
+        note << " here:";
+      }
+  }
+  if (!reportInto)
+    LLVM_DEBUG(llvm::dbgs() << "  - Visited " << *expr << ": " << ineq << "\n");
+
+  return ineq;
 }
 
 /// Solve the constraint problem. This is a very simple implementation that
 /// does not fully solve the problem if there are weird dependency cycles
 /// present.
-void ConstraintSolver::solve() {
+LogicalResult ConstraintSolver::solve() {
+  // Ensure that there are no adverse cycles around.
+  LLVM_DEBUG(llvm::dbgs() << "Checking for unbreakable loops\n");
+  SmallPtrSet<Expr *, 16> seenVars;
+  bool anyFailed = false;
+
+  for (auto *expr : exprs) {
+    // Only work on variables.
+    auto *var = dyn_cast<VarExpr>(expr);
+    if (!var)
+      continue;
+    LLVM_DEBUG(llvm::dbgs()
+               << "- Checking " << *var << " >= " << *var->constraint << "\n");
+
+    // Canonicalize the variable's constraint expression into a form that allows
+    // us to easily determine if any recursion leads to an unsatisfiable
+    // constraint. The `seenVars` set acts as a recursion breaker.
+    seenVars.insert(var);
+    auto ineq = checkCycles(var, var->constraint, seenVars, info);
+    seenVars.clear();
+
+    // If the constraint is satisfiable, we're done.
+    // TODO: It's possible that this result is already sufficient to arrive at a
+    // solution for the constraint, and the second pass further down is not
+    // necessary. This would require more proper handling of `MinExpr` in the
+    // cycle checking code.
+    if (ineq.sat()) {
+      LLVM_DEBUG(llvm::dbgs()
+                 << "  - Breakable since " << ineq << " satisfiable\n");
+      continue;
+    }
+
+    // If we arrive here, the constraint is not satisfiable at all. To provide
+    // some guidance to the user, we call the cycle checking code again, but
+    // this time with an in-flight diagnostic to attach notes indicating
+    // unsatisfiable paths in the cycle.
+    LLVM_DEBUG(llvm::dbgs()
+               << "  - UNBREAKABLE since " << ineq << " unsatisfiable\n");
+    anyFailed = true;
+    for (auto value : info.find(var)->second) {
+      // Depending on whether this value stems from an operation or not, create
+      // an appropriate diagnostic identifying the value.
+      auto op = value.getDefiningOp();
+      auto diag =
+          op ? op->emitOpError() : mlir::emitError(value.getLoc()) << "value ";
+      diag << "is constrained to be wider than itself";
+
+      // Re-run the cycle checking, but this time reporting into the diagnostic.
+      seenVars.insert(var);
+      checkCycles(var, var->constraint, seenVars, info, &diag);
+      seenVars.clear();
+    }
+  }
+
+  // If there were unbreakable cycles, there's no sense in inferring any widths.
```

Comment:
Probably, but for debugging there might be value in inferring as many widths as possible and leave the uninferred ones alone, so that dumping the ir after the failed pass carries more information.

Reply:
Definitely not required for this PR.

Reply:
You're right, and the solver actually works in the presence of adverse cycles (it may just leave solutions `None`). I dropped this early exit to improve debuggability :+1:.